### DISCUSSION TITLE
Общее обсуждение #4 // 2024

### DISCUSSION THREAD

**Discussion by Kor-Angar** [dosinabox/g2nr_unofficial_update#537] [General]
....

**Comment by Kor-Angar** [comment]
у тебя есть проверки в скриптах для НПС магов . она идёт в ряде скриптов. я думаю можно их вынести в отдельный скрипт что бы это править в одном месте.

**Reply by Kor-Angar** [reply]
проверка для людей магов используется в заклах

- сон
- страх
- забыть(?)

**Reply by Kor-Angar** [reply]
люди маги это:
* маги огня 
* маги воды
* ксардас
* ватрас
* миксир(?)
* Daron

**Reply by Kor-Angar** [reply]
* палы?

**Reply by Kor-Angar** [reply]
сон
***
	if(Hlp_GetInstanceID(npc) == Hlp_GetInstanceID(Daron))
	{
		return FALSE;
	};
	if(Hlp_GetInstanceID(npc) == Hlp_GetInstanceID(Vatras))
	{
		return FALSE;
	};
	if(npc.guild == GIL_KDF)
	{
		return FALSE;
	};
	if(npc.guild == GIL_PAL)
	{
		return FALSE;
	};
	if(npc.guild == GIL_DMT)
	{
		return FALSE;
	};
	if(npc.guild == GIL_KDW)
	{
		return FALSE;
	};
	return TRUE;


страх
****
```D
if((self.guild == GIL_KDF) 
    || (self.guild == GIL_PAL) 
    || (self.guild == GIL_KDW) 
    || (Hlp_GetInstanceID(self) == Hlp_GetInstanceID(Xardas)) 
    || (Hlp_GetInstanceID(self) == Hlp_GetInstanceID(Andre)) 
    || C_IsNpc(self,MIL_304_Torwache) 
    || (Hlp_GetInstanceID(self) == Hlp_GetInstanceID(Torwache_305)) 
    || (Hlp_GetInstanceID(self) == Hlp_GetInstanceID(Vatras)) 
    || (Hlp_GetInstanceID(self) == Hlp_GetInstanceID(Myxir_CITY)) 
    || (Hlp_GetInstanceID(self) == Hlp_GetInstanceID(Daron)))
```

**Reply by Kor-Angar** [reply]
черновик
```D
// маги
if (self.guild == GIL_KDF) {	return FALSE; };
if  (self.guild == GIL_KDW) {	return FALSE; }; 
if  (Hlp_GetInstanceID(self) == Hlp_GetInstanceID(Xardas)) {	return FALSE; }; 
if (Hlp_GetInstanceID(self) == Hlp_GetInstanceID(Vatras))  {	return FALSE; };
if  (Hlp_GetInstanceID(self) == Hlp_GetInstanceID(Myxir_CITY))  {	return FALSE; };
if (Hlp_GetInstanceID(self) == Hlp_GetInstanceID(Daron))) {	return FALSE; };


// палы
if (self.guild == GIL_PAL) {	return FALSE; }; 
if  (Hlp_GetInstanceID(self) == Hlp_GetInstanceID(Andre))  {	return FALSE; };
if (Hlp_GetInstanceID(self) == Hlp_GetInstanceID(Torwache_305))  {	return FALSE; };
if C_IsNpc(self,MIL_304_Torwache)   {	return FALSE; };

return TRUE;
```

**Reply by dosinabox** [reply]
Ну да. Получилось удобно. В процессе вскрылось, например, что не все паладины и маги лечились магией: https://github.com/dosinabox/g2nr_unofficial_update/commit/2b7a2c45870839b1095f194e9bf667074eb6a9dd

**Reply by Kor-Angar** [reply]
C_NpcIsMage. навернео С_НПСИзХуманМаг так как  есть ещё ПНСмаги монстры.

**Reply by Kor-Angar** [reply]
а чего так хитро то? ;-)
```D
	if(C_IsNpc(slf,MIL_304_Torwache))
	{
		return TRUE;
	};
	if(Hlp_GetInstanceID(slf) == Hlp_GetInstanceID(Torwache_305))
	{
		return TRUE;
	};
```

**Reply by Kor-Angar** [reply]
в этой проверки смысла нет  `&& (self.attribute[ATR_MANA] >= SPL_Cost_PalFullHeal))`, так как в скриптах мана у НПС пополняется перед кастом

**Reply by Kor-Angar** [reply]
к проверки по AIV_MagicUser  в [ZS_HealSelf.d](https://github.com/dosinabox/g2nr_unofficial_update/commit/2b7a2c45870839b1095f194e9bf667074eb6a9dd#diff-8d191048f62358a1ad7d7d6933b93f3e3e2af3dbc50f918092aad8b3e84c044d) тоже можно было бы присмотреться. 

![image](https://github.com/dosinabox/g2nr_unofficial_update/assets/49514764/053e06ea-f307-40d7-b0c4-72ea58d76815)

**Reply by Kor-Angar** [reply]
> к проверки по AIV_MagicUser тоже можно было бы присмотреться.


в [ZS_HealSelf.d](https://github.com/dosinabox/g2nr_unofficial_update/commit/2b7a2c45870839b1095f194e9bf667074eb6a9dd#diff-8d191048f62358a1ad7d7d6933b93f3e3e2af3dbc50f918092aad8b3e84c044d)

**Reply by dosinabox** [reply]
> C_NpcIsMage. навернео С_НПСИзХуманМаг так как есть ещё ПНСмаги монстры.

Ага. C_NpcIsMonsterMage.

> а чего так хитро то? ;-)

Глобального вара для 304 нету, и ради одной проверки добавлять не хочется.

**Reply by Kor-Angar** [reply]
так сделай обоим проверку по инстанции ;-)

**Comment by Kor-Angar** [comment]
а ты Лорду Андре гильдию так и не поменяла? :-)

**Reply by dosinabox** [reply]
нет

**Comment by dosinabox** [comment]
Неиспользуемое свойство TrueGuild можно использовать для хранения новых "второстепенных" гильдий ("работники борделя", "воры", "беглецы Моргахарда", "жители верхнего квартала"). Это позволит при необходимости их идентифицировать без перебора по инстанциям и, например, научить заступаться друг за друга.

**Reply by dosinabox** [reply]
После консультаций с Граттом: лучше не трогать TrueGuild вообще, они проверяются в движке.
А для "подгильдий" можно использовать новый aivar, который заменит собой существующие:

- AIV_NPCIsRanger
- AIV_StoryBandit
- AIV_StoryBandit_Esteban

**Reply by Kor-Angar** [reply]
в aivar[43] можно записать.

**Reply by Kor-Angar** [reply]
и сразу бы подумать для подгильдий. на вскидку я предлагаю подгильдии для BAU на фермах Хориноса.

бау_лобарт
бау_акил
бау_секоб
бау_бенгар
бау_онар

тоже самое с паладинами. разделить паладинов по гео

**Comment by Kor-Angar** [comment]
такую проверку можно использовать в B_AssignAmbientInfos. 

```D
func void B_AssignAmbientInfos(var C_NPC slf)
{
    // НПС что то хочет сказать герою
    if (Npc_CheckInfo (self, 1))    { return; }
    // Герой что то хочет спросить у НПС
    if (Npc_CheckInfo (self, 0))    { return; }
        
    // с НПС нет индивидуальных разговоров. Подключаем  шаблоны
```

**Comment by Kor-Angar** [comment]
при создание сета со шпагой B_CreateAmbientInv_BDT было бы хорошо добавить проверку на уровень, что бы не создавать ItMw_Addon_BanditTrader для низкоуровневых бандитов

`if self.level > 3`

**Reply by dosinabox** [reply]
Есть такое в планах. И дело не в уровне бандитов, а в мире: эти шпаги продает бандитам Фернандо, и они должны быть только в Хоринисе (и в Яркендаре?). 
Проблема в том, что на уровне инстанции (откуда вызывается функция создания инвентаря) нельзя проверять мир.

**Reply by Kor-Angar** [reply]
> Проблема в том, что на уровне инстанции (откуда вызывается функция создания инвентаря) нельзя проверять мир.

проверку можно сделать. при старте игры(Хоринис) глава будет 0 и CurrentLevel будет 0. после загрузки мира переменная в CurrentLevel будет уже записана цифра. 

соответственно в STARTUP_OldWorld  / STARTUP_AddonWorld  переменная CurrentLevel будет больше 0.

***

то что я писал выше касается низкоуровневых бандитов у которых не будет хватать параметров что бы одеть данную шпагу.

**Reply by Kor-Angar** [reply]
в принципе можно и переменную добавить  STARTUPLevel  для того что бы понимать какой запускается STARTUP

**Comment by Kor-Angar** [comment]
а разве DIA_DiegoNW_NeedHelp_WhoAreYou может быть доступно? ведь разговор во 2ой главе с диего обязателен.

**Reply by Kor-Angar** [reply]
можно попробовать дополнить первый диалог с диего в минентале.


```D
func void DIA_DiegoOw_Hallo_End()
{
	AI_Output(self,other,"DIA_DiegoOw_Hallo_11_06");	//Я хотел предупредить паладинов о драконах. Сам не знаю, зачем мне это было нужно.
	AI_Output(self,other,"DIA_DiegoOw_Hallo_11_07");	//Этот напыщенный Лотар даже не стал слушать меня - и, конечно же, не позволил мне поговорить с лордом Хагеном.
	AI_Output(self,other,"DIA_DiegoOw_Hallo_11_08");	//Вместо этого они отправили меня назад, с экспедицией. Так что лучше не напоминай мне о паладинах...
	AI_Output(other,self,"DIA_DiegoOw_Hallo_15_09");	//Очень важно остановить драконов, пока у нас еще есть время - и не важно, кто поможет нам сделать это.
	AI_Output(self,other,"DIA_DiegoOw_Hallo_11_10");	//Остановить их? Да нам нужно уносить ноги из этой долины, пока у нас еще есть время для этого!
	AI_Output(self,other,"DIA_DiegoOw_Silvestro_11_03");	//Скажи - как тебе удалось пробраться через Проход? Я думал, там все кишит орками.
	if(PlayerPassedAbandonedMine == TRUE)
	{
		AI_Output(other,self,"DIA_DiegoOw_Silvestro_15_04");	//Есть путь через заброшенные шахты, где нет орков.
		AI_Output(self,other,"DIA_DiegoOw_Silvestro_11_05");	//Это хорошая новость. Тогда я должен как можно быстрее отправляться в Хоринис - мне там нужно кое с кем свести счеты.
	}
	else
	{
		B_PC_Thief_JustLucky();
	};
};



func void DIA_DiegoOw_Hallo_WhoAreYou()
{
	AI_Output(other,self,"DIA_DiegoNW_NeedHelp_WhoAreYou_15_00");	//Кто ты?
	//AI_Output(self,other,"DIA_DiegoNW_NeedHelp_WhoAreYou_11_01");	//Черт, это все, наверное, из-за одежды. В старой одежде стражники не пустили бы меня в город.
	//AI_Output(self,other,"DIA_DiegoNW_NeedHelp_WhoAreYou_11_02");	//Вот почему я купил это одеяние у торговца за городом. Теперь-то, надеюсь, ты вспомнил меня? Я Диего.
	AI_Output(self,other,"DIA_DiegoNW_NeedHelp_WhoAreYou_11_03");	//В старой колонии я учил тебя всему, что необходимо для выживания.
	AI_Output(self,other,"DIA_DiegoNW_NeedHelp_WhoAreYou_11_04");	//Ты же не мог вот так просто все забыть.
	AI_Output(other,self,"DIA_DiegoNW_NeedHelp_WhoAreYou_15_05");	//Ээ... а что именно произошло там?
	AI_Output(self,other,"DIA_DiegoNW_NeedHelp_WhoAreYou_11_06");	//Черт побери! Ты действительно ничего не помнишь.
	AI_Output(self,other,"DIA_DiegoNW_NeedHelp_WhoAreYou_11_07");	//Послушай, у меня действительно нет времени объяснять все тебе теперь. Послушай только одно: мы были чертовски хорошими друзьями, и я не раз спасал твою жизнь.
	//AI_Output(self,other,"DIA_DiegoNW_NeedHelp_WhoAreYou_11_08");	//А теперь мне нужна твоя помощь.
    // подобрать какую то реплику ГГ 
    DIA_DiegoOw_Hallo_End();
};

func void DIA_DiegoOw_Hallo_Hagen()
{
	AI_Output(other,self,"DIA_DiegoOw_Hallo_15_01");	//Да... так было.
	AI_Output(other,self,"DIA_DiegoOw_Hallo_15_02");	//Ну, теперь я вернулся и ищу доказательства появления драконов.
	AI_Output(self,other,"DIA_DiegoOw_Hallo_11_03");	//Для кого ты это делаешь?
	AI_Output(other,self,"DIA_DiegoOw_Hallo_15_04");	//Я работаю на лорда Хагена. С помощью паладинов драконов можно остановить.
	AI_Output(self,other,"DIA_DiegoOw_Hallo_11_05");	//Паладины? Позволь мне сказать тебе кое-что. После того, как мне удалось вырваться отсюда, я отправился в Хоринис.
    DIA_DiegoOw_Hallo_End();
};



func void DIA_DiegoOw_Hallo_Info()
{
	AI_Output(self,other,"DIA_DiegoOw_Hallo_11_00");	//Эй, я думал, ты мертв.
    Info_ClearChoices(DIA_DiegoOw_Hallo);
    Info_AddChoice(DIA_DiegoOw_Hallo,"Кто ты?",         DIA_DiegoOw_Hallo_WhoAreYou);
    Info_AddChoice(DIA_DiegoOw_Hallo,"Да... так было.", DIA_DiegoOw_Hallo_Hagen);
};
```

**Reply by dosinabox** [reply]
В обновлении разговор с Диего не обязателен. Вся нужная информация есть в записке на трупе Сильвестро.

**Reply by Kor-Angar** [reply]
записка подразумевает намерение. встреча с диего свершённое событие. как же герой может отдать записку и сказать что Диего с рудой там и там то не зная этого?  особенно учитывая что Диего еще и конвой перебил :-D

**Reply by dosinabox** [reply]
Если встречи с Диего не было, то в диалоге с Гарондом Диего и не упоминается. Сейчас не могу посмотреть подробнее.

**Reply by Kor-Angar** [reply]
дело не в этом. дело в том что задание Гаронда узнать где и сколько руды. а без разговора с Диего об этом узнать нельзя. то есть разговор с Диего обязателен. такая же ситуация со старой шахтой, герой тоже туда может прийти и узнать что Маркос переместил руду в другое место. но Гаронд требует отчёт.
